File: src/main/scala/database/wallet/SecureWalletDatabase.java
package database.wallet;

import java.io.File;

import org.mapdb.DB;
import org.mapdb.DBMaker;
import org.mapdb.Serializer;

import scorex.account.PrivateKeyAccount;
import settings.Settings;

public class SecureWalletDatabase 
{
	private static final File SECURE_WALLET_FILE = new File(Settings.walletDir(), "wallet.s.dat");
	
	private static final String SEED = "seed";
	private static final String NONCE = "nonce";
	
	private DB database;
	
	private AccountSeedMap accountSeedMap;
	
	public static boolean exists()
	{
		return SECURE_WALLET_FILE.exists();
	}
	
	public SecureWalletDatabase(String password)
	{
		//OPEN WALLET
		SECURE_WALLET_FILE.getParentFile().mkdirs();
				
		//DELETE TRANSACTIONS
		//File transactionFile = new File(Settings.getInstance().getWalletDir(), "wallet.s.dat.t");
		//transactionFile.delete();	
		
		this.database = DBMaker.newFileDB(SECURE_WALLET_FILE)
						.encryptionEnable(password)
			    		.closeOnJvmShutdown()
			    		.cacheSize(2048)
			    		.checksumEnable()
			    		.mmapFileEnableIfSupported()
			            .make();
			    
		this.accountSeedMap = new AccountSeedMap(this, this.database);
	}
	
	public AccountSeedMap getAccountSeedMap()
	{
		return this.accountSeedMap;
	}

	public void setSeed(byte[] seed) 
	{
		this.database.createAtomicVar(SEED, seed, Serializer.BYTE_ARRAY);
	}
	
	public byte[] getSeed()
	{
		return (byte[]) this.database.getAtomicVar(SEED).get();
	}

	public void setNonce(int nonce) 
	{
		 this.database.getAtomicInteger(NONCE).set(nonce);	
	}
	
	public int getNonce()
	{
		return this.database.getAtomicInteger(NONCE).intValue();
	}
	
	public int getAndIncrementNonce()
	{
		return this.database.getAtomicInteger(NONCE).getAndIncrement();
	}
	
	public void delete(PrivateKeyAccount account)
	{
		this.accountSeedMap.delete(account);
	}
	
	public void commit()
	{
		this.database.commit();
	}
	
	public void close() 
	{
		if(this.database != null)
		{
			if(!this.database.isClosed())
			{
				this.database.commit();
				this.database.close();
			}
		}
	}
}

File: src/main/scala/database/wallet/SecureWalletDatabase.scala
package database.wallet

import java.io.File

import org.mapdb.{DBMaker, Serializer}
import scorex.account.PrivateKeyAccount
import settings.Settings

import scala.collection.JavaConversions._


class SecureWalletDatabase(password: String) {

  import SecureWalletDatabase._

  //OPEN WALLET
  SECURE_WALLET_FILE.getParentFile.mkdirs()

  private val database = DBMaker.newFileDB(SECURE_WALLET_FILE)
    .encryptionEnable(password)
    .closeOnJvmShutdown()
    .cacheSize(2048)
    .checksumEnable()
    .mmapFileEnableIfSupported()
    .make()

  private val accountsMap = database.createHashMap("accounts").makeOrGet[String, PrivateKeyAccount]()

  def addAccount(account: PrivateKeyAccount) = {
    val address = account.address
    if (!accountsMap.containsKey(address)) {
      accountsMap.put(address, account)
      database.commit()
      true
    } else false
  }

  def accounts() = accountsMap.values().toSeq

  def account(address: String) = Option(accountsMap.get(address))

  def setSeed(seed: Array[Byte]) = database.createAtomicVar(SEED, seed, Serializer.BYTE_ARRAY)

  def getSeed(): Array[Byte] = database.getAtomicVar(SEED).get()

  def getNonce(): Int = database.getAtomicInteger(NONCE).intValue()

  def setNonce(nonce: Int) = database.getAtomicInteger(NONCE).set(nonce)

  def getAndIncrementNonce(): Int = database.getAtomicInteger(NONCE).getAndIncrement()

  def delete(account: PrivateKeyAccount) {
    accountsMap.remove(account.address)
  }

  def commit() = database.commit()

  def close() {
    if (!database.isClosed) {
      database.commit()
      database.close()
    }
  }
}

object SecureWalletDatabase {
  private val SECURE_WALLET_FILE = new File(Settings.walletDir, "wallet.s.dat")
  private val SEED = "seed"
  private val NONCE = "nonce"

  def exists() = SECURE_WALLET_FILE.exists()
}